abe/ma-abe_test.go

```diff
@@ -0,0 +1,172 @@
+    // create new MAABE struct with Global Parameters
+    maabe := abe.NewMAABE()
+
+    // create three authorities, each with two attributes
+    attribs1 := []string{"auth1:at1", "auth1:at2"}
+    attribs2 := []string{"auth2:at1", "auth2:at2"}
+    attribs3 := []string{"auth3:at1", "auth3:at2"}
+    auth1, err:= maabe.NewMAABEAuth("auth1", attribs1)
+    if err != nil {
+        t.Fatalf("Failed generation authority %s: %v\n", "auth1", err)
+    }
+    auth2, err:= maabe.NewMAABEAuth("auth2", attribs2)
+    if err != nil {
+        t.Fatalf("Failed generation authority %s: %v\n", "auth2", err)
+    }
+    auth3, err:= maabe.NewMAABEAuth("auth3", attribs3)
+    if err != nil {
+        t.Fatalf("Failed generation authority %s: %v\n", "auth3", err)
+    }
+
+    // create a msp struct out of the boolean formula
+    msp, err := abe.BooleanToMSP("((auth1:at1 AND auth2:at1) OR (auth1:at2 AND auth2:at2)) OR (auth3:at1 AND auth3:at2)", false)
+    if err != nil {
+        t.Fatalf("Failed to generate the policy: %v\n", err)
+    }
+
+    // define the set of all public keys we use
+    pks := []*abe.MAABEPubKey{auth1.Pk, auth2.Pk, auth3.Pk}
+
+    // choose a message
+    msg := "Attack at dawn!"
+
+    // encrypt the message with the decryption policy in msp
+    ct, err := maabe.Encrypt(msg, msp, pks)
+    if err != nil {
+        t.Fatalf("Failed to encrypt: %v\n", err)
+    }
+
+    // also check for empty message
+    msg_empty := ""
+    _, err = maabe.Encrypt(msg_empty, msp, pks)
+    assert.Error(t, err)
+
+    // use a pub keyring that is too small
+    pks_small := []*abe.MAABEPubKey{auth1.Pk}
+    _, err = maabe.Encrypt(msg, msp, pks_small)
+    assert.Error(t, err)
+
+    // choose a single user's Global ID
+    gid := "gid1"
+
+    // authority 1 issues keys to user
+    key11, err := auth1.GenerateAttribKey(gid, "auth1:at1", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth1:at1", err)
+    }
+    key12, err := auth1.GenerateAttribKey(gid, "auth1:at2", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth1:at2", err)
+    }
+    // authority 2 issues keys to user
+    key21, err := auth2.GenerateAttribKey(gid, "auth2:at1", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth2:at1", err)
+    }
+    key22, err := auth2.GenerateAttribKey(gid, "auth2:at2", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth2:at2", err)
+    }
+    // authority 3 issues keys to user
+    key31, err := auth3.GenerateAttribKey(gid, "auth3:at1", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth3:at1", err)
+    }
+    key32, err := auth3.GenerateAttribKey(gid, "auth3:at2", maabe)
+    if err != nil {
+        t.Fatalf("Failed to generate attribute key for %s: %v\n", "auth3:at2", err)
+    }
```

Comment:
I think that it would be nicer if there was function `GenerateAttribKeys` that would accept a list of attributes and return a list of keys, instead of individual calls.